pkg/sql/catalog/table_elements.go

```diff
@@ -408,30 +433,153 @@ type Constraint interface {
 	// GetConstraintValidity returns the validity of this constraint.
 	GetConstraintValidity() descpb.ConstraintValidity
 
+	// IsEnforced returns true iff the constraint is enforced for all writes to
+	// the parent table.
+	IsEnforced() bool
+
 	// GetName returns the name of this constraint update mutation.
 	GetName() string
 
-	// NotNullColumnID returns the underlying not-null column ID, if there is one.
-	NotNullColumnID() descpb.ColumnID
+	// IsConstraintValidated returns true iff the constraint is enforced for
+	// all writes to the table data and has also been validated on the table's
+	// existing data prior to the addition of the constraint, if there was any.
+	IsConstraintValidated() bool
+
+	// IsConstraintUnvalidated returns true iff the constraint is enforced for
+	// all writes to the table data but which is explicitly NOT to be validated
+	// on any data in the table prior to the addition of the constraint.
+	IsConstraintUnvalidated() bool
+}
+
+// UniqueConstraint is an interface for a unique constraint.
+type UniqueConstraint interface {
+	Constraint
+
+	// IsValidReferencedUniqueConstraint returns whether the unique constraint can
+	// serve as a referenced unique constraint for a foreign key constraint.
+	IsValidReferencedUniqueConstraint(fk ForeignKeyConstraint) bool
+
+	// NumKeyColumns returns the number of columns in this unique constraint.
+	NumKeyColumns() int
+
+	// GetKeyColumnID returns the ID of the column in the unique constraint at
+	// ordinal `columnOrdinal`.
+	GetKeyColumnID(columnOrdinal int) descpb.ColumnID
+
+	// CollectKeyColumnIDs returns the columns in the unique constraint in a new
+	// TableColSet.
+	CollectKeyColumnIDs() TableColSet
+
+	// IsPartial returns true iff this is a partial index.
+	IsPartial() bool
+
+	// GetPredicate returns the partial predicate if there is one, "" otherwise.
+	GetPredicate() string
+}
+
+// UniqueWithIndexConstraint is an interface around a unique constraint
+// which backed by an index.
+type UniqueWithIndexConstraint interface {
+	UniqueConstraint
+	Index
```

Comment:
Just a comment: If we use this interface to represent a PK or UNIQUE constraint, and we want to ask the question whether it's a PK or UNIQUE, we will need to look at `obj.GetEncodingType` instead of `obj.IsPrimary`, both of which are methods under `catalog.Index` interface.

Reply:
I've deliberately avoided distinguishing PK constraints from the more general UNIQUE case. I haven't found it to be necessary to do so _yet_ but that doesn't mean that it won't be in the future.